### PR TITLE
fix_46_allow_for_missing_actions_while_rendering

### PR DESCRIPTION
**Pull request by nanuxbe**
Hey @carltongibson ,

here's the (initial?) PR for #46 with the changes discussed in the issue.

While creating this PR though, I noticed it was rather straightforward to **remove** actions but that adding an action (ie: duplicate, archive, open, ...) would be rather complicated due to the nature of `Role` that cannot be extended.

I am thinking about refactoring `Role` to be more flexible (either not an `Enum` anymore or implementing something like a `role_factory`). It would probably be best left in its own PR but it could very well be included here too.

### CODE REVIEW

**Comment by carltongibson** [comment]
> While creating this PR though, I noticed it was rather straightforward to remove actions but that adding an action (ie: duplicate, archive, open, ...) would be rather complicated due to the nature of Role that cannot be extended.
>
> I am thinking about refactoring Role to be more flexible (either not an Enum anymore or implementing something like a role_factory). It would probably be best left in its own PR but it could very well be included here too.

Ah, yes. Separate issue. 😅 I have thoughts in this direction, that I need to upstream. Roughly: The core CRUD roles should be an enum — there just are the options there are. BUT we need a role protocol, which can be implemented by a custom role-like composing the core CRUD roles plus additional extra actions. 

At that point it's feasible that we can automatically add links to the templates for custom actions.

File: src/neapolitan/templatetags/neapolitan.py
Comment (carltongibson) [inline src/neapolitan/templatetags/neapolitan.py:16]: Is this sufficient? 🤔

If I do `BookmarksCRUDView.get_urls(roles={Role.LIST})` will not these not raise a reverse match error? 

Could you add a test case for that example, asserting that the `bookmark-list` view renders correctly, and that these links are absent? (That should cover both the `get_context_data()` case and this one here.)

Comment (nanuxbe) [reply src/neapolitan/templatetags/neapolitan.py:16]: They will not raise an error because of [this change](https://github.com/carltongibson/neapolitan/pull/47/files#diff-15775856a482440af9196b06d3627395adc2f09fae171621cb2cbddb08fa3463R91) which catches the `NoReverseMatch` exception and returns `None` instead.

I'll add a test.

Something I did on other projects was to have an extendable test case that tests if all possible actions render with a satisfactory response code (>=200,<400). But I think that should probably wait until ROLE protocol is a thing

Comment (carltongibson) [reply src/neapolitan/templatetags/neapolitan.py:16]: > They will not raise an error because of [this change](https://github.com/carltongibson/neapolitan/pull/47/files#diff-15775856a482440af9196b06d3627395adc2f09fae171621cb2cbddb08fa3463R91)... 

Ah, yes. I didn't spot that. Let me consider this. 🤔 (Is it changing the contract of `reverse` too much — often you're going to **want** `NoReverseMatch` when you call the reverse for a non-routed handler )

> I'll add a test.

Great. Thanks. 

> Something I did on other projects ...

Yes! A CRUDViewTestCase kind of thing is on the radar to give base coverage. (But yes, 😜, pending resolving the custom roles issue.)

**Comment by dacodekid** [comment]
@nanuxbe, I just tried your branch, but still getting the same `-create' is not a valid view function or pattern name.` error.

```toml
# pyproject.toml
neapolitan = {git = "https://github.com/nanuxbe/neapolitan.git"}
```

**Comment by carltongibson** [comment]
@dacodekid You probably need to specify the PR branch.

**Comment by dacodekid** [comment]
@carltongibson 

I tried that as well but still getting the same error.  I'll wait until this gets merged.

```toml
neapolitan = { git = "https://github.com/nanuxbe/neapolitan.git", branch = "fix_#46_NoReverseMatch" }
```

**Review by carltongibson — COMMENTED** [review]
Hey @nanuxbe — I pushed a few edits in dadc4b3e59d985a411e0671ddc7a9a62cde5bc87. Let me know what you think.

**Review by nanuxbe — COMMENTED** [review]
LGTM @carltongibson , I like the maybe_reverse approach!

**Comment by carltongibson** [comment]
Great. Thanks for the follow up @nanuxbe. I'll pull this in later on. Have a great Sunday.

**Review by carltongibson — APPROVED** [review]

**Pull request merged**
